pages/chaoss/visualizations/contrib_importance_over_time.py

```diff
@@ -0,0 +1,478 @@
+        Input(f"window-width-{PAGE}-{VIZ_ID}", "value"),
+        Input(f"step-size-{PAGE}-{VIZ_ID}", "value"),
+        Input(f"date-picker-range-{PAGE}-{VIZ_ID}", "start_date"),
+        Input(f"date-picker-range-{PAGE}-{VIZ_ID}", "end_date"),
+    ],
+    background=True,
+)
+
+def create_contrib_prolificacy_over_time_graph(repolist, patterns, threshold, window_width, step_size, start_date, end_date):
+    # main function for all data pre processing
+    cache = cm()
+    df = cache.grabm(func=ctq, repos=repolist)
+    
+    while df is None:
+        time.sleep(1.0)
+        df = cache.grabm(func=ctq, repos=repolist)
+
+    # data ready.
+    start = time.perf_counter()
+    logging.warning(f"{VIZ_ID}- START")
+
+    # test if there is data
+    if df.empty:
+        logging.warning(f"{VIZ_ID} - NO DATA AVAILABLE")
+        return nodata_graph
+    
+    # if the step size is greater than window width raise Alert
+    if step_size > window_width:
+        return dash.no_update, True
+    
+    df_final = process_data(df, 
+                            patterns, 
+                            threshold, 
+                            window_width, 
+                            step_size, 
+                            start_date, 
+                            end_date)
+
+    fig = create_figure(df_final, step_size)
+
+    logging.warning(f"{VIZ_ID} - END - {time.perf_counter() - start}")
+    return fig, False
+
+def calc_cntrb_prolificacy(df, action_type, threshold):
+    # if the df is empty return None
+    if df.empty:
+        return None
+    
+    # if the specified action type is not in the dfs' cols return None
+    if action_type not in df.columns:
+        return None 
+    
+    # sort rows in df based on number of contributions from greatest to least
+    df = df.sort_values(by=action_type, ascending=False)
+
+    # calculate the threshold amount of contributions 
+    thresh_cntrbs = df[action_type].sum() * threshold
+    
+    # initilize running sum of contributors who make up contributor prolificacy
+    cntrb_prolificacy = 0
+
+    # initialize running sum of contributions
+    running_sum = 0
+    
+    for _, row in df.iterrows():
+        running_sum += row[action_type] # update the running sum by the number of contributions a contributor has made
+        cntrb_prolificacy+=1 # update contributor prolificacy 
+        # if the running sum of contributions is greater than or equal to the threshold amount, break
+        if running_sum >= thresh_cntrbs:
+            break
+
+    return cntrb_prolificacy
+
+def cntrb_prolificacy_over_time(df, period_from, period_to, window_width, threshold):
+    # subset df such that the rows correspond to the window of time defined by period from and period to
+    time_mask = (df['created_at'] >= period_from) & (df['created_at'] <= period_to)
+    df_in_range = df.loc[time_mask]
+    
+    # drop rows where the cntrib_id is None
+    cntrb_mask = (df_in_range['cntrb_id'] == None)
```

Comment:
By doing num here it is making the thresh_cntrbs in calc_cntrb_prolificacy incorrect as it doesnt include the contributions from contributors that are nulls